CI fails for slow tests: NotImplementedError: "_amp_foreach_non_finite_check_and_unscale_cuda" not implemented for 'BFloat16'

CI fails for slow tests: https://github.com/huggingface/trl/actions/runs/21370086538/job/61512072745
> NotImplementedError: "_amp_foreach_non_finite_check_and_unscale_cuda" not implemented for 'BFloat16'
```python
FAILED tests/test_sft_trainer.py::TestSFTTrainerSlow::test_sft_trainer_transformers_mp[trl-internal-testing/tiny-LlamaForCausalLM-3.2-True] - NotImplementedError: "_amp_foreach_non_finite_check_and_unscale_cuda" not implemented for 'BFloat16'
FAILED tests/test_sft_trainer.py::TestSFTTrainerSlow::test_sft_trainer_transformers_mp[trl-internal-testing/tiny-LlamaForCausalLM-3.2-False] - NotImplementedError: "_amp_foreach_non_finite_check_and_unscale_cuda" not implemented for 'BFloat16'
FAILED tests/test_sft_trainer.py::TestSFTTrainerSlow::test_sft_trainer_transformers_mp[trl-internal-testing/tiny-MistralForCausalLM-0.2-True] - NotImplementedError: "_amp_foreach_non_finite_check_and_unscale_cuda" not implemented for 'BFloat16'
FAILED tests/test_sft_trainer.py::TestSFTTrainerSlow::test_sft_trainer_transformers_mp[trl-internal-testing/tiny-MistralForCausalLM-0.2-False] - NotImplementedError: "_amp_foreach_non_finite_check_and_unscale_cuda" not implemented for 'BFloat16'
FAILED tests/test_sft_trainer.py::TestSFTTrainerSlow::test_sft_trainer_transformers_mp_gc[trl-internal-testing/tiny-LlamaForCausalLM-3.2-True-None] - NotImplementedError: "_amp_foreach_non_finite_check_and_unscale_cuda" not implemented for 'BFloat16'
FAILED tests/test_sft_trainer.py::TestSFTTrainerSlow::test_sft_trainer_transformers_mp_gc[trl-internal-testing/tiny-LlamaForCausalLM-3.2-True-gradient_checkpointing_kwargs1] - NotImplementedError: "_amp_foreach_non_finite_check_and_unscale_cuda" not implemented for 'BFloat16'
FAILED tests/test_sft_trainer.py::TestSFTTrainerSlow::test_sft_trainer_transformers_mp_gc[trl-internal-testing/tiny-LlamaForCausalLM-3.2-True-gradient_checkpointing_kwargs2] - NotImplementedError: "_amp_foreach_non_finite_check_and_unscale_cuda" not implemented for 'BFloat16'
FAILED tests/test_sft_trainer.py::TestSFTTrainerSlow::test_sft_trainer_transformers_mp_gc[trl-internal-testing/tiny-LlamaForCausalLM-3.2-False-None] - NotImplementedError: "_amp_foreach_non_finite_check_and_unscale_cuda" not implemented for 'BFloat16'
FAILED tests/test_sft_trainer.py::TestSFTTrainerSlow::test_sft_trainer_transformers_mp_gc[trl-internal-testing/tiny-LlamaForCausalLM-3.2-False-gradient_checkpointing_kwargs1] - NotImplementedError: "_amp_foreach_non_finite_check_and_unscale_cuda" not implemented for 'BFloat16'
FAILED tests/test_sft_trainer.py::TestSFTTrainerSlow::test_sft_trainer_transformers_mp_gc[trl-internal-testing/tiny-LlamaForCausalLM-3.2-False-gradient_checkpointing_kwargs2] - NotImplementedError: "_amp_foreach_non_finite_check_and_unscale_cuda" not implemented for 'BFloat16'
FAILED tests/test_sft_trainer.py::TestSFTTrainerSlow::test_sft_trainer_transformers_mp_gc[trl-internal-testing/tiny-MistralForCausalLM-0.2-True-None] - NotImplementedError: "_amp_foreach_non_finite_check_and_unscale_cuda" not implemented for 'BFloat16'
FAILED tests/test_sft_trainer.py::TestSFTTrainerSlow::test_sft_trainer_transformers_mp_gc[trl-internal-testing/tiny-MistralForCausalLM-0.2-True-gradient_checkpointing_kwargs1] - NotImplementedError: "_amp_foreach_non_finite_check_and_unscale_cuda" not implemented for 'BFloat16'
FAILED tests/test_sft_trainer.py::TestSFTTrainerSlow::test_sft_trainer_transformers_mp_gc[trl-internal-testing/tiny-MistralForCausalLM-0.2-True-gradient_checkpointing_kwargs2] - NotImplementedError: "_amp_foreach_non_finite_check_and_unscale_cuda" not implemented for 'BFloat16'
FAILED tests/test_sft_trainer.py::TestSFTTrainerSlow::test_sft_trainer_transformers_mp_gc[trl-internal-testing/tiny-MistralForCausalLM-0.2-False-None] - NotImplementedError: "_amp_foreach_non_finite_check_and_unscale_cuda" not implemented for 'BFloat16'
FAILED tests/test_sft_trainer.py::TestSFTTrainerSlow::test_sft_trainer_transformers_mp_gc[trl-internal-testing/tiny-MistralForCausalLM-0.2-False-gradient_checkpointing_kwargs1] - NotImplementedError: "_amp_foreach_non_finite_check_and_unscale_cuda" not implemented for 'BFloat16'
FAILED tests/test_sft_trainer.py::TestSFTTrainerSlow::test_sft_trainer_transformers_mp_gc[trl-internal-testing/tiny-MistralForCausalLM-0.2-False-gradient_checkpointing_kwargs2] - NotImplementedError: "_amp_foreach_non_finite_check_and_unscale_cuda" not implemented for 'BFloat16'
```

Traceback:
```python
>       trainer.train()

tests/test_sft_trainer.py:1862: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
.venv/lib/python3.11/site-packages/transformers/trainer.py:2174: in train
    return inner_training_loop(
.venv/lib/python3.11/site-packages/transformers/trainer.py:2569: in _inner_training_loop
    _grad_norm = self.accelerator.clip_grad_norm_(
.venv/lib/python3.11/site-packages/accelerate/accelerator.py:3008: in clip_grad_norm_
    self.unscale_gradients()
.venv/lib/python3.11/site-packages/accelerate/accelerator.py:2946: in unscale_gradients
    self.scaler.unscale_(opt)
.venv/lib/python3.11/site-packages/torch/amp/grad_scaler.py:343: in unscale_
    optimizer_state["found_inf_per_device"] = self._unscale_grads_(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <torch.amp.grad_scaler.GradScaler object at 0x7fcdc5a39010>
optimizer = AdamW (
Parameter Group 0
    amsgrad: False
    betas: (0.9, 0.999)
    capturable: False
    decoupled_weight_decay:...1e-08
    foreach: None
    fused: True
    initial_lr: 2e-05
    lr: 2e-05
    maximize: False
    weight_decay: 0.0
)
inv_scale = tensor(1.5259e-05, device='cuda:0')
found_inf = tensor(0., device='cuda:0'), allow_fp16 = False

    def _unscale_grads_(
        self,
        optimizer: torch.optim.Optimizer,
        inv_scale: torch.Tensor,
        found_inf: torch.Tensor,
        allow_fp16: bool,
    ) -> dict[torch.device, torch.Tensor]:
        per_device_inv_scale = _MultiDeviceReplicator(inv_scale)
        per_device_found_inf = _MultiDeviceReplicator(found_inf)
    
        # To set up _amp_foreach_non_finite_check_and_unscale_, split grads by device and dtype.
        # There could be hundreds of grads, so we'd like to iterate through them just once.
        # However, we don't know their devices or dtypes in advance.
    
        # https://stackoverflow.com/questions/5029934/defaultdict-of-defaultdict
        # Google says mypy struggles with defaultdicts type annotations.
        per_device_and_dtype_grads: dict[
            torch.device, dict[torch.dtype, list[torch.Tensor]]
        ] = defaultdict(lambda: defaultdict(list))
        with torch.no_grad():
            for group in optimizer.param_groups:
                for param in group["params"]:
                    assert isinstance(param, torch.Tensor)
                    if param.grad is None:
                        continue
                    if (not allow_fp16) and param.grad.dtype == torch.float16:
                        raise ValueError("Attempting to unscale FP16 gradients.")
                    if param.grad.is_sparse:
                        # is_coalesced() == False means the sparse grad has values with duplicate indices.
                        # coalesce() deduplicates indices and adds all values that have the same index.
                        # For scaled fp16 values, there's a good chance coalescing will cause overflow,
                        # so we should check the coalesced _values().
                        if param.grad.dtype is torch.float16:
                            param.grad = param.grad.coalesce()
                        to_unscale = param.grad._values()
                    else:
                        to_unscale = param.grad
    
                    # TODO: is there a way to split by device and dtype without appending in the inner loop?
                    per_device_and_dtype_grads[to_unscale.device][
                        to_unscale.dtype
                    ].append(to_unscale)
    
            for device, per_dtype_grads in per_device_and_dtype_grads.items():
                for grads in per_dtype_grads.values():
>                   torch._amp_foreach_non_finite_check_and_unscale_(
                        grads,
                        per_device_found_inf.get(device),
                        per_device_inv_scale.get(device),
                    )
E                   NotImplementedError: "_amp_foreach_non_finite_check_and_unscale_cuda" not implemented for 'BFloat16'

.venv/lib/python3.11/site-packages/torch/amp/grad_scaler.py:280: NotImplementedError
```

--- a/tests/test_sft_trainer.py
+++ b/tests/test_sft_trainer.py
@@ -1848,7 +1848,7 @@ class TestSFTTrainerSlow(TrlTestCase):
             max_length=self.max_length,
         )
 
-        model = AutoModelForCausalLM.from_pretrained(model_name)
+        model = AutoModelForCausalLM.from_pretrained(model_name, dtype="float32")
         tokenizer = AutoTokenizer.from_pretrained(model_name)
 
         trainer = SFTTrainer(
@@ -1892,7 +1892,7 @@ class TestSFTTrainerSlow(TrlTestCase):
             gradient_checkpointing_kwargs=gradient_checkpointing_kwargs,
         )
 
-        model = AutoModelForCausalLM.from_pretrained(model_name)
+        model = AutoModelForCausalLM.from_pretrained(model_name, dtype="float32")
         tokenizer = AutoTokenizer.from_pretrained(model_name)
 
         trainer = SFTTrainer(

